### PR TITLE
Add document for CA1851

### PR DESCRIPTION
**Pull request by Cosifne**
Hi,
I checked in a new analyzer in Roslyn Analyzer repo 
[PR in roslyn-analyzer repo](https://github.com/dotnet/roslyn-analyzers/pull/5907)
[Design in the runtime repo](https://github.com/dotnet/runtime/issues/63381)
And this is the document for it

### CODE REVIEW

**Review by Youssef1313 — CHANGES_REQUESTED** [review]
This needs an entry in [performance-warnings.md](https://github.com/dotnet/docs/blob/main/docs/fundamentals/code-analysis/quality-rules/performance-warnings.md) as well.

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
+---
+title: "CA1851: Possible multiple enumerations of `IEnumerable` collection"
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:2]: I don't think titles and descriptions should use backticks.

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+| **Fix is breaking or non-breaking** |Non-breaking|
+
+## Cause
+
+Multiple enumerations of `IEnumerable` collections are detected.
+
+## Rule description
+
+Collection with [IEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerable?view=net-6.0) or [IEnumerable\<T\>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-6.0) type generated by many LINQ methods like [Select](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.select?view=net-6.0) and [yield return](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/yield) statements has the ability to defer enumeration when it is generated. The enumeration will start as long as it is passed into enumeration LINQ methods like [ElementAt](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.elementat?view=net-6.0) or used in [for each statement](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements#the-foreach-statement). The enumeration result is not calculated once and cached like [Lazy](https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1?view=net-6.0).
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:31]: All these links should be `xref`s

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+public void MyMethod(IEnumerable<int> input)
+{
+    // If the underlying type of 'input' is List<int>
+    var inputArray = (List<int>)input;
+    var count = inputArray.Count();
+    foreach (var i in inputArray) { }
+}
+```
+
+If the underlying type of `IEnumerable` collection is an iterator-based implementation generated by [yield return](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/yield) statements. It still can be fixed by converting and caching the collection to other types to fix the diagnostic. But it needs extra memory.
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:64]: This should be a relative link.

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    // and it takes O(n) time to finish, where 'n' is the length of 'someStrings'.
+    var someStringsArray = someStrings.ToArray()
+
+    // It takes 2 * O(1) to complete 'Count()' and 'Last()' calls.
+    var count = someStringsArray.Count();
+    var lastElement = someStringsArray.Last();
+}
+```
+
+## Configure customized enumeration methods and LINQ chain methods.
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:98]: Titles shouldn't end in periods

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+By default, all the customized methods accepting `IEnumerable` parameters are considered not enumerating the parameters. It can changed by setting the [`assume_method_enumerates_parameters`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#assume-method-enumerates-parameters) in editorconfig.
+
+## When to suppress warnings
+
+It is safe to suppress the warnings if the underlying of the `IEnumerable` collection is some other type like `List` or `Array`, or if you are sure the methods accepting `IEnumerable` collection would not enumerated it.
+
+## See also
+
+[Deferred execution example (LINQ to XML)](https://docs.microsoft.com/en-us/dotnet/standard/linq/deferred-execution-example)
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:114]: Should be relative link

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+**Violation**:
+
+```csharp
+public void MyMethod(IEnumerable<int> input)
+{
+    var count = inputArray.Count();
+    foreach (var i in inputArray) { }
+}
+```
```
Comment (Youssef1313) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:50]: There need to be VB code snippets I think.

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    var count = someStringsArray.Count();
+    var lastElement = someStringsArray.Last();
+}
+```
+
+## Configure customized enumeration methods and LINQ chain methods.
+
+By default, all the methods under `System.Linq` namespace are included in the analysis scope. Customized methods enumerating the `IEnumerable` parameters could be added into the scope by setting the [`enumeration_methods`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#enumeration-methods) in editorconfig.
+
+Customized LINQ chain methods (methods take `IEnumerable` parameters and return a new `IEnumerable` instance) could also be added by setting the [`linq_chain_methods`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#linq-chain-methods) in editorconfig.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:102]:
```suggestion
You can also add customized LINQ chain methods (that is, methods take an `IEnumerable` argument and return a new `IEnumerable` instance) to the analysis scope by setting the [`linq_chain_methods`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#linq-chain-methods) option in an *.editorconfig* file.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+## Configure the default assumption of methods take `IEnumerable` parameters
+
+By default, all the customized methods accepting `IEnumerable` parameters are considered not enumerating the parameters. It can changed by setting the [`assume_method_enumerates_parameters`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#assume-method-enumerates-parameters) in editorconfig.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:106]:
```suggestion
By default, all customized methods that accept an `IEnumerable` argument are assumed not to enumerate the argument. You can change this by setting the [`assume_method_enumerates_parameters`](https://github.com/dotnet/roslyn-analyzers/blob/main/docs/Analyzer%20Configuration.md#assume-method-enumerates-parameters) option in an *.editorconfig* file.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1851.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+## When to suppress warnings
+
+It is safe to suppress the warnings if the underlying of the `IEnumerable` collection is some other type like `List` or `Array`, or if you are sure the methods accepting `IEnumerable` collection would not enumerated it.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1851.md:110]:
```suggestion
It is safe to suppress this warning if the underlying of the `IEnumerable` collection is some other type like `List` or `Array`, or if you're sure that methods that take an `IEnumerable` collection do not enumerate it.
```

File: docs/fundamentals/code-analysis/quality-rules/index.md [hunk 10]
```diff
@@ -144,6 +144,7 @@ The following table lists code quality analysis rules.
 > | [CA1848: Use the LoggerMessage delegates](ca1848.md) | For improved performance, use the `LoggerMessage` delegates. |
 > | [CA1849: Call async methods when in an async method](ca1849.md) | In a method which is already asynchronous, calls to other methods should be to their async versions, where they exist. |
 > | [CA1850: Prefer static `HashData` method over `ComputeHash`](ca1850.md) | It's more efficient to use the static `HashData` method over creating and managing a `HashAlgorithm` instance to call `ComputeHash`. |
+> | [CA1851: Possible multiple enumerations of `IEnumerable` collection](ca1851.md) | Possible multiple enumerations of `IEnumerable` collection. Consider using an implementation that avoid multiple enumerations. |
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/index.md:147]:
```suggestion
> | [CA1851: Possible multiple enumerations of `IEnumerable` collection](ca1851.md) | Possible multiple enumerations of `IEnumerable` collection. Consider using an implementation that avoids multiple enumerations. |
```

**Review by gewarren — COMMENTED** [review]
Thanks for adding this!

**Comment by Cosifne** [comment]
Ah, I forget this PR in the last few weeks, will update it soon.

**Review by gewarren — APPROVED** [review]

**Pull request merged**